src/overlay/TCPPeer.h

```diff
@@ -26,15 +28,80 @@ class TCPPeer : public Peer
     static constexpr size_t BUFSZ = 0x40000; // 256KB
 
   private:
+    // Helper class which provides invariance for various data structures;
+    // Each method should be called from the same thread
+    class ThreadRestrictedVars
+    {
+        std::deque<TimestampedMessage> mWriteQueue;
+        std::vector<asio::const_buffer> mWriteBuffers;
+        bool const mUseBackgroundThread;
+        bool mWriting{false};
+        bool mSocketShutdownScheduled{false};
+        std::vector<uint8_t> mIncomingHeader;
+        std::vector<uint8_t> mIncomingBody;
+
+      public:
+        ThreadRestrictedVars(bool useBackgroundThread)
+            : mUseBackgroundThread(useBackgroundThread)
+        {
+        }
+        std::deque<TimestampedMessage>&
+        getWriteQueue()
+        {
+            releaseAssert(!threadIsMain() || !mUseBackgroundThread);
+            return mWriteQueue;
+        }
+        std::vector<asio::const_buffer>&
+        getWriteBuffers()
+        {
+            releaseAssert(!threadIsMain() || !mUseBackgroundThread);
+            return mWriteBuffers;
+        }
+        std::vector<uint8_t>&
+        getIncomingHeader()
+        {
+            releaseAssert(!threadIsMain() || !mUseBackgroundThread);
+            return mIncomingHeader;
+        }
+        std::vector<uint8_t>&
+        getIncomingBody()
+        {
+            releaseAssert(!threadIsMain() || !mUseBackgroundThread);
+            return mIncomingBody;
+        }
+        bool
+        isWriting() const
+        {
+            releaseAssert(!threadIsMain() || !mUseBackgroundThread);
+            return mWriting;
+        }
+        void
+        setWriting(bool value)
+        {
+            releaseAssert(!threadIsMain() || !mUseBackgroundThread);
+            mWriting = value;
+        }
+        bool
+        socketShutdownScheduled() const
+        {
+            releaseAssert(threadIsMain());
+            return mSocketShutdownScheduled;
+        }
+        void
+        scheduleSocketShutdown(bool value)
```

Comment:
Remove `bool value`. This should only every be called once anyway and `true` is the only valid value, so remove parameter.